### DISCUSSION TITLE
auth.uid() in triggers

### DISCUSSION THREAD

**Discussion by magick93** [supabase/supabase#9066] [Questions] [answered]
I'm trying to use the `auth.uid()` function within a trigger.

The trigger is 

```
CREATE OR REPLACE FUNCTION public.claim_template_updated()
 RETURNS trigger
 LANGUAGE plpgsql
AS $function$
        DECLARE
        user_id             uuid            := auth.uid();
    BEGIN
      NEW.updated_at        = now();
      NEW.updated_by_user   = auth.uid();
      RETURN NEW;
    END;
    $function$
;
```

I'm able to use the `auth.uid()` in a function, called using `rpc` and it works as expected. But not in a trigger. 

Is this a known limitation, or a bug, or am I doing something wrong?

Thanks

**Comment by GaryAustin1** [comment]
That should work as long as the trigger is being fired by a user from the API.   
What are you seeing as far as error or not working?

**Reply by magick93** [reply]
May be I'm missing something obvious. My `update` trigger is working correctly but `insert` is not.

**Insert trigger function - not storing the user_id**

```
create trigger claim_template_inserted before
insert
    on
    public.claim_template for each row execute function claim_template_inserted()

CREATE OR REPLACE FUNCTION public.claim_template_inserted()
 RETURNS trigger
 LANGUAGE plpgsql
AS $function$
        DECLARE
        user_id             uuid            := auth.uid();
    BEGIN
      NEW.created_at        = now();
      NEW.created_by_user   = auth.uid();
      RETURN NEW;
    END;
    $function$
;
```

**Update trigger - working as expected**

```
create trigger claim_template_updated before
update
    on
    public.claim_template for each row execute function claim_template_updated()


CREATE OR REPLACE FUNCTION public.claim_template_updated()
 RETURNS trigger
 LANGUAGE plpgsql
AS $function$
        DECLARE
        user_id             uuid            := auth.uid();
    BEGIN
      NEW.updated_at        = now();
      NEW.updated_by_user   = auth.uid();
      RETURN NEW;
    END;
    $function$
;

```

**Reply by GaryAustin1** [reply]
Unless you have user_id defaulted in the table, or passed in (on update I assume you do), you are not setting it here.  
You need to do NEW.user_id = auth.uid().

**Reply by magick93** [reply]
I dont have a column named `user_id`. Initially I tried to `DECLARE` a variable called `user_id` and set the value using `auth.uid()`. 

The column name that I'm trying to set, which has a relationship to `auth.user_id` is `created_by_user`.

**Reply by magick93** [reply]
I've also tried setting the `created_by_user` column default to `auth.uid()` but it still doesnt get set.

**Comment by magick93** [comment, marked as answer]
Found the issue. 

It was due to a session being cleared in Cypress tests - not related to supabase / postgres triggers.
